File: 0433-minimum-genetic-mutation/0433-minimum-genetic-mutation.py
from typing import List
from collections import deque

class Solution:
    def minMutation(self, startGene: str, endGene: str, bank: List[str]) -> int:
        genes = set(bank)
        if endGene not in genes:
            return -1
        q = deque([startGene])
        res = 0
        dna = ['A', 'C', 'G', 'T']
        while q:
            for _ in range(len(q)):
                gene = q.popleft()
                if gene == endGene:
                    return res
                for i in range(len(gene)):
                    for c in dna:
                        tmp = gene[:i] + c + gene[i+1:]
                        if tmp in genes:
                            q.append(tmp)
                            genes.remove(tmp)
            res += 1
        return -1

File: 0433-minimum-genetic-mutation/0433-minimum-genetic-mutation.cpp
class Solution {
public:
    int minMutation(string startGene, string endGene, vector<string>& bank) {
        unordered_set<string> genes(bank.begin(), bank.end());
        if (!genes.count(endGene)) return -1;
        queue<string> q{{startGene}};
        int res = 0;
        vector<char> dna{'A', 'C', 'G', 'T'};
        while (!q.empty()) {
            for (int k = q.size(); k > 0; --k) {
                auto gene = q.front(); q.pop();
                if (gene == endGene) return res;
                for (int i = 0; i < gene.size(); ++i) {
                    auto tmp = gene;
                    for (auto c : dna) {
                        tmp[i] = c;
                        if (genes.count(tmp)) {
                            q.emplace(tmp);
                            genes.erase(tmp);
                        }
                    }
                }
            }
            ++res;
        }
        return -1;
    }
};
